include/envoy/network/filter.h

```diff
@@ -147,6 +148,67 @@ class FilterManager {
   virtual bool initializeReadFilters() PURE;
 };
 
+/**
+ * Callbacks used by individual listener filter instances to communicate with the listener filter
+ * manager.
+ */
+class ListenerFilterCallbacks {
+public:
+  virtual ~ListenerFilterCallbacks() {}
+
+  /**
+   * @return the AcceptSocket that owns this manager and the managed filters.
+   */
+  virtual AcceptSocket& socket() PURE;
+
+  /**
+   * @return the Dispatcher for this manager.
+   */
+  virtual Event::Dispatcher& dispatcher() PURE;
+
+  /**
+   * If an filter stopped filter iteration by returning FilterStatus::StopIteration,
+   * the filter should call continueFilterChain(true) when complete to continue the filter chain,
+   * or continueFilterChain(false) if the filter execution failed and the connection musrt be
+   * closed.
+   * @param success boolean telling whether the filter execution was successful or not.
+   */
+  virtual void continueFilterChain(bool success) PURE;
+};
+
+/**
+ * Listener Filter
+ */
+class ListenerFilter {
+public:
+  virtual ~ListenerFilter() {}
+
+  /**
+   * Called when a new connection is accepted, but before a Connection is created.
+   * Filter chain iteration can be stopped if needed.
+   * @param cb the callbacks the filter instance can use to communicate with the filter chain.
+   * @return status used by the filter manager to manage further filter iteration.
+   */
+  virtual FilterStatus onAccept(ListenerFilterCallbacks& cb) PURE;
+};
+
+typedef std::shared_ptr<ListenerFilter> ListenerFilterSharedPtr;
```

Comment:
I think this can probably be unique_ptr. The reason the other filters are shared_ptr is because of read/write/both semantics. That doesn't apply here.

Reply:
Will try this out.

Reply:
This was complicated a bit due to google mock not working with lvalue references. Worked around it by passing a raw pointer to addAcceptFilter() method.